src/dir.rs

```diff
@@ -0,0 +1,167 @@
+///!    * can be iterated through multiple times without closing and reopening the file
+///!      descriptor. Each iteration rewinds when finished.
+///!    * returns entries for `.` (current directory) and `..` (parent directory).
+///!    * returns entries' names as a `CStr` (no allocation or conversion beyond whatever libc
+///!      does).
+pub struct Dir(ptr::NonNull<libc::DIR>);
+
+impl Dir {
+    #[inline]
+    pub fn from<F: IntoRawFd>(fd: F) -> Result<Self> {
+        Dir::from_fd(fd.into_raw_fd())
+    }
+
+    /// Converts from a file descriptor, closing it on success or failure.
+    pub fn from_fd(fd: RawFd) -> Result<Self> {
+        match ptr::NonNull::new(unsafe { libc::fdopendir(fd) }) {
+            None => {
+                let e = Error::last();
+                unsafe { libc::close(fd) };
+                Err(e)
+            }
+            Some(d) => Ok(Dir(d)),
+        }
+    }
+}
+
+// `Dir` is not `Sync`, but it's not reference-counted either, so it can be safely passed from one
+// thread to another.
+unsafe impl Send for Dir {}
+
+impl<'d> IntoIterator for &'d mut Dir {
+    type Item = Result<Entry>;
+    type IntoIter = Iter<'d>;
+    fn into_iter(self) -> Self::IntoIter {
+        Iter(self)
+    }
+}
+
+impl AsRawFd for Dir {
+    fn as_raw_fd(&self) -> RawFd {
+        unsafe { libc::dirfd(self.0.as_ptr()) }
+    }
+}
+
+impl fmt::Debug for Dir {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Dir")
+            .field("fd", &self.as_raw_fd())
+            .finish()
+    }
+}
+
+impl Drop for Dir {
+    fn drop(&mut self) {
+        unsafe { libc::closedir(self.0.as_ptr()) };
+    }
+}
+
+#[allow(missing_debug_implementations)]
+pub struct Iter<'d>(&'d mut Dir);
+
+impl<'d> Iterator for Iter<'d> {
+    type Item = Result<Entry>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        unsafe {
+            // Note: POSIX specifies that portable applications should dynamically allocate a
+            // buffer with room for a `d_name` field of size `pathconf(..., _PC_NAME_MAX)` plus 1
+            // for the NUL byte. It doesn't look like the std library does this; it just uses
+            // fixed-sized buffers (and libc's dirent seems to be sized so this is appropriate).
+            // Probably fine here too then.
+            let mut ent: Entry = Entry(::std::mem::uninitialized());
+            let mut result = ptr::null_mut();
+            if let Err(e) = Errno::result(readdir_r((self.0).0.as_ptr(), &mut ent.0, &mut result)) {
+                return Some(Err(e));
+            }
+            if result == ptr::null_mut() {
+                return None;
+            }
+            assert_eq!(result, &mut ent.0 as *mut dirent);
+            return Some(Ok(ent));
+        }
+    }
+}
+
+impl<'d> Drop for Iter<'d> {
+    fn drop(&mut self) {
+        unsafe { libc::rewinddir((self.0).0.as_ptr()) }
```

Comment:
Are the parens around `self.0` really necessary?

Reply:
Yeah, I put them here at the suggestion of the compiler:

```
error: unexpected token: `0.0`
  --> src/dir.rs:97:58
   |
97 |             if let Err(e) = Errno::result(readdir_r(self.0.0.as_ptr(), &mut ent.0, &mut result)) {
   |                                                     -----^^^
   |                                                     |    |
   |                                                     |    unexpected token
   |                                                     help: try parenthesizing the first index: `(self.0).0`
```